### ISSUE TITLE
Please make the Paho MQTT types implement `std::marker::Send`

### ISSUE THREAD

**Issue by berndv** [eclipse/paho.mqtt.rust#31]
**Description**
I would like to write an application that accepts incoming requests, calculates and returns the result. The calculation takes a long time, so several calculations should run in parallel and do not block other incoming requests.

A simple solution is to start a worker thread for each incoming request, which performs the computation and publishes the result when the calculation has finished.

Unfortunately, it is not possible to implement this simple pattern with Phao MQTT because the Phao types do not implement the marker trait `std::marker::Send`.

Note: For some reasons, I do want to use the synchronous API in my real life project.

**My Request**
Please make the Phao types implement `std::marker::Send`.

**Code Example**
Here's a snippet that illustrates the problem (rustc 1.30.1):

```
extern crate paho_mqtt;

use paho_mqtt::{Client, Message, MessageBuilder};
use std::sync::{Arc, Mutex};
use std::thread;
use std::time::Duration;

fn main() {
    // create client and establish connection to broker
    let mut client = Client::new("tcp://localhost:1883").unwrap();
    client.connect(None).unwrap();

    // prepare to consume incoming requests
    let requests = client.start_consuming();
    client.subscribe("requests", 1).unwrap();

    // move client into mutex to share it with other threads
    let client = Arc::new(Mutex::new(client));

    // dispatch incoming requests, start a worker thread per request
    loop {
        let request = requests.recv().unwrap().unwrap();
        // create a thread private clone of the client protected by the mutex
        let client_clone = client.clone();
        thread::spawn(move || {    // <--- Error: "`*mut std::ffi::c_void` cannot be sent between threads safely" and more...
            // some long running operation
            let response = process_request(request);
            // publish the result
            client_clone.lock().unwrap().publish(response).unwrap();
        });
    }
}

fn process_request(request: Message) -> Message {
    thread::sleep(Duration::from_millis(3000));
    MessageBuilder::new()
        .topic("responses")
        .payload("42")
        .qos(1)
        .finalize()
}
```
Part of the build output:

```
foo@ubuntu:~/git/mqtt-send$ cargo build
   Compiling mqtt-send v0.1.0 (/home/foo/git/mqtt-send)
error[E0277]: `*mut std::ffi::c_void` cannot be sent between threads safely
  --> src/main.rs:25:9
   |
25 |         thread::spawn(move || {    // <--- Error: "`*mut std::ffi::c_void` cannot be sent between threads safely" and more...
   |         ^^^^^^^^^^^^^ `*mut std::ffi::c_void` cannot be sent between threads safely
   |
   = help: within `[closure@src/main.rs:25:23: 30:10 request:paho_mqtt::Message, client_clone:std::sync::Arc<std::sync::Mutex<paho_mqtt::Client>>]`, the trait `std::marker::Send` is not implemented for `*mut std::ffi::c_void`
   = note: required because it appears within the type `paho_mqtt_sys::MQTTAsync_message`
   = note: required because it appears within the type `paho_mqtt::Message`
   = note: required because it appears within the type `[closure@src/main.rs:25:23: 30:10 request:paho_mqtt::Message, client_clone:std::sync::Arc<std::sync::Mutex<paho_mqtt::Client>>]`
   = note: required by `std::thread::spawn`

error[E0277]: `*mut paho_mqtt_sys::MQTTClient_persistence` cannot be sent between threads safely
  --> src/main.rs:25:9
...
```

**Comment by fpagliughi** [comment]
Already working on it!  Should have something pushed to the 'develop' branch in about a week.

**Comment by janwilts** [comment]
Is there any update on this?

**Comment by fpagliughi** [comment]
Oh, yeah, wow. This has been a pretty long week! 
(Looking at my last comment)
Work has been a distraction recently, but I mean to get back on this soon to push toward MQTT v5.
I will have a look at the progress I made on this in January, and try to push it ASAP.

**Comment by fpagliughi** [comment]
OK. They types are marked as _Send_, and sometime _Sync_ as appropriate. I just pushed this to the _develop_ branch, and will likely make a v0.6 release, dare I say it, in about a week!

Apologies for the delay. I would love any feedback on this.

The ironic thing is that the types were thread-safe from the beginning. The clients and Token have interior mutability, but they are protected by Mutex'es, since they manage callbacks from a separate thread in the C library. So really, all I needed to do all this time was to mark the structs as Send, and maybe Sync, where appropriate.

So it comes out:
**AsyncClient** and **Client** -> _Send_ and _Sync_ (Internally protected with mutexes)
**Token** -> _Send_ (also internal mutex)
Most option structs -> _Send_ and _Sync_ (Raw pointers always point into the structs themselves, never aliased)

The **Token** is likely _Sync_ as well, but the intent was never to have multiple threads waiting for a single Token, even though it was possible. Now that they are Futures, I need to look closer into them, to see if that's still true.

Note though, that my though for the clients was that they might need to be mutable during creation and initialization, but after that, most communication can be done through immutable references. So, in most normal cases, you might not need a Mutex, but just an Arc for the client:

```
let cli = AsyncClient::new("tcp://localhost:1883").unwrap();

let cli = Arc::new(cli);
let cli2 = cli.clone();

cli.connect().wait();

let thr = thread::spawn(move || {
    assert!(cli.is_connected());
    // ...
});
assert(cli2.is_connected());
// ...
```

**Comment by fpagliughi** [comment]
This does make me reconsider one other thing, though...

Both the _AsyncClient_ and _Token_ structs are thin wrappers around heap-allocated "inner" structs. That is because we pass pointers to these inner structs to the C library and don't want them moving in memory (see Issue #17).

The Token uses an Arc to the inner struct, like:
```
pub struct Token {
    inner: Arc<TokenInner>,
}
```

whereas the client uses a Box. This allows us to clone a Token easily, which just clones the Arc (so all clones of a Token refer to the same inner data). But the same is not true for the client. We would need to put the client in an external Arc to share it.

This was intentional, to restrict central control of the client. But if that's out the window now, we may reconsider these implementation details.
 
For example, the Token has a raw pointer back to the client. This would seem to violate lifetimes, but the pointer is only used during a callback from the client, so it's guaranteed to be alive at least until the end of the callback.

So, if we use Arc's for the inner client, we may consider updating some aspects of the callbacks and API to make use of this.

**Comment by fpagliughi** [comment]
Fix went into the develop branch at 565ea527de983bde2cd9e4b696bb639e391ee6b4. It will go out with the v0.6 release shortly.
